scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -1516,6 +1536,152 @@
                  next-considerable))))))
 
+(defn job->task-txn
+  "Generate task transaction for a job."
+  [job metadata compute-cluster]
+  (let [{:keys [job/last-waiting-start-time job/uuid]} job
+        {:keys [executor task-id]} metadata
+        job-ref [:job/uuid uuid]
+        instance-start-time (now)]
+    [[:job/allowed-to-start? job-ref]
+     ; The job will remain waiting until the instance is running.
+     [:db/add job-ref :job/state :job.state/waiting]
+     (cond->
+      {:db/id (d/tempid :db.part/user)
+       :job/_instance job-ref
+       :instance/executor executor
+       :instance/executor-id task-id ; NB command executor uses the task-id as the executor-id 
+       :instance/hostname "Unknown"
+       :instance/ports 9876
+       :instance/preempted? false
+       :instance/progress 0
+       :instance/slave-id "Unknown"
+       :instance/start-time instance-start-time
+       :instance/status :instance.status/unknown
+       :instance/task-id task-id
+       :instance/compute-cluster (cc/db-id compute-cluster)}
+       last-waiting-start-time
+       (assoc :instance/queue-time
+              (- (.getTime instance-start-time)
+                 (.getTime ^Date last-waiting-start-time))))]))
+
+(defn jobs->kubernetes-task-metadata
+  "Generate sequence of task-metadata for jobs in a compute-cluster, suitable for Kubernetes Scheduler."
+  [jobs pool-name mesos-run-as-user compute-cluster]
+  (let [jobs->task-id (plumbing.core/map-from-keys (fn [_] (str (d/squuid))) jobs)]
+    (for [{:keys [job/name job/user job/uuid job/environment] :as job} jobs]
+      (let [pool-specific-resources
+            ((adjust-job-resources-for-pool-fn pool-name) job (tools/job-ent->resources job))]
+        (merge
+         (task/job->task-metadata compute-cluster mesos-run-as-user
+                                  job (jobs->task-id job))
+         {; NB: see `kubernetes/compute_cluster/autoscale!`
+          ; for how launching a real job pod without a hostname
+          ; compares to launching a synthetic pod.
+
+          ; Job constraints need to be expressed on a real job
+          ; pod so that if it triggers the cluster autoscaler,
+          ; CA will spin up nodes that will end up satisfying it. 
+          :pod-constraints (constraints/job->constraints job)
+
+          ; We need to label the pods handled completely by Kubernetes
+          ; so that we can treat them differently throughout the scheduler.
+          :pod-labels {"twosigma.com/scheduler" "kubernetes"}
+
+          ; Cook has a "novel host constraint", which disallows a job from
+          ; running on the same host twice. So, we need to avoid running a
+          ; real job pod on any of the hosts that it won't be able to run on.
+          :pod-hostnames-to-avoid (constraints/job->previous-hosts-to-avoid job)
+          :task-request {:scalar-requests (walk/stringify-keys pool-specific-resources)
+                         :job {:job/pool {:pool/name pool-name}
```

Comment:
Why rebuild this from the components? Can the `job` be used directly, instead of destructuring and rebuilding a new one? (And if not, can you explain why in a comment?)

Reply:
Good catch. I was providing just enough to run but was missing providing `job/application` as an example.